api/src/main/java/org/eclipse/microprofile/metrics/MetricRegistry.java

```diff
@@ -298,52 +300,131 @@ static String name(Class<?> klass, String... names) {
+     * @return a new or pre-existing {@link Gauge}
      *
      * @since 3.0
      */
-    Gauge<?> gauge(String name, Gauge<?> gauge, Tag...tags);
+    <T> Gauge<Long> gauge(MetricID metricID, T object, ToDoubleFunction<T> func);
 
     /**
-     * Return the {@link Gauge} registered under the {@link MetricID}; or create and register this
-     * gauge if none is registered.
+     * Return the {@link Gauge} of type Long registered under the {@link MetricID} with the @{link Metadata}'s
+     * name and with the provided {@link Tag}s; or create and register this gauge if none is registered.
      *
      * If a {@link Gauge} was created, a {@link Metadata} object will be registered with the name
      * and type. If a {@link Metadata} object is already registered with this metric name then that
      * {@link Metadata} will be used.
      *
-     * @param metricID the ID of the metric
-     * @param gauge the {@link Gauge} to use if none is registered already
-     * @return the pre-existing or provided {@link Gauge}
+     * The created {@link Gauge} will apply a @link java.util.function.ToDoubleFunction ToDoubleFunction}
+     * to the provided object to resolve a Long value.
+     *
+     * @param <T>      The Type of the Object of which the function <code>func</code> is applied to
+     * @param metadata The Metadata of the Gauge
+     * @param object   The object that the {@link java.util.function.ToDoubleFunction ToDoubleFunction} <code>func</code> will be applied to
+     * @param func     The {@link java.util.function.ToDoubleFunction ToDoubleFunction} that will be applied to <code>object</code>
+     * @param tags     The tags of the metric
+     * @return a new or pre-existing {@link Gauge}
+     *
+     * @since 3.0
+     */
+    <T> Gauge<Long> gauge(Metadata metadata, T object, ToDoubleFunction<T> func, Tag... tags);
+
+    /**
+     * Return the {@link Gauge} registered under the {@link MetricID} with this
+     * name and with the provided {@link Tag}s; or create and register this gauge if none is registered.
+     *
+     * If a {@link Gauge} was created, a {@link Metadata} object will be registered with the name
+     * and type. If a {@link Metadata} object is already registered with this metric name then that
+     * {@link Metadata} will be used.
+     *
+     * The created {@link Gauge} will return the value that the {@link java.util.function.Supplier Supplier}
+     * will provide.
+     *
+     * @param <T>      The type that {@link Gauge} metric will return
+     * @param name     The name of the Gauge
+     * @param supplier The {@link java.util.function.Supplier Supplier} function that will return the value for the Gauge metric
+     * @param tags     The tags of the metric
+     * @return a new or pre-existing {@link Gauge}
+     *
+     * @since 3.0
+     */
+    <T> Gauge<T> gauge(String name, Supplier<T> supplier, Tag... tags);
```

Comment:
This can be Supplier<Number>... you don't need to do any cast on the supplier function (as long as it returns some kind of number, you don't need to cast anything). Again, that returned Gauge is just for reference.. the output of the Supplier is what is observed/measured.

Reply:
Yes but we've had these problems in the past that while gauges were designed to return only numbers, the API allows registering a gauge that returns anything. You can register a `Gauge<String>` and unless the implementation proactively checks that, then it will most likely break during export.
So here the user could call this with a `Supplier<String>` and break stuff with it.